### DISCUSSION TITLE
Extending PackageRules by GroupName

### DISCUSSION THREAD

**Discussion by danhlee329** [renovatebot/renovate#28183] [Request Help]
### What would you like help with?

I would like help with my configuration

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

Github

### Please tell us more about your question or problem

Moving the question from https://github.com/renovatebot/renovate/discussions/27865#discussioncomment-8934220, reaching out to see if this is possible as I am confused at this point.  Is it possible to extend a particular `packageRule` by the `groupName`?

For example, say we have the config-in-repo (see below) and we want to extend a particular package rule in the config-in-app (see below).  Will the merged config be what's captured below (under `Example Config Extended (Final merged config)`)? Looks like [excludePackageNames](https://docs.renovatebot.com/configuration-options/#excludepackagenames) is `mergable`, but how for a particular package rule?

Also, if we pass in an empty array or null for `excludePackageNames` when we extend a config, would that be a no-op (i.e., no changes made to the package rule at all)?

Part of the confusion is that when testing things out with the extended config, the debug logs show the `combined config` without showing the changes made with the referenced config (see example in the original reply above).

<details><summary>Example Config (Repo)</summary>

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [],
  "packageRules": [
    {
      "groupName": ["test-group-name"],
      "matchDatasources": ["custom.local_module_tracker"],
      "branchName": "renovate/local_module_tracker",
      "labels": ["renovate", "local_module_tracker"]
    }
  ],
  "hostRules": [],
  "requireConfig": "required",
  "platform": "github",
  "token": "***********"
}
```

</details>

<details><summary>Example Config Extended (App)</summary>

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": ["github>$ORG/$REPO//renovate.json"],
  "packageRules": [
    {
      "groupName": ["test-group-name"],
      "excludePackageNames": ["dep_name"]
    }
  ],
  "hostRules": [],
  "requireConfig": "required",
  "platform": "github",
  "token": "***********"
}
```

</details>

<details><summary>Example Config Extended (Final merged config)</summary>

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": ["github>$ORG/$REPO//renovate.json"],
  "packageRules": [
    {
      "groupName": ["test-group-name"],
      "matchDatasources": ["custom.local_module_tracker"],
      "branchName": "renovate/local_module_tracker",
      "labels": ["renovate", "local_module_tracker"],
      "excludePackageNames": ["dep_name"]
    }
  ],
  "hostRules": [],
  "requireConfig": "required",
  "platform": "github",
  "token": "***********"
}
```

</details>

**Comment by danhlee329** [comment]
I guess `matchDatasources` would be sufficient for us too, just curious how Renovate knows what to merge and what not to?

**Reply by viceice** [reply]
depends on the mergable flag, which is also shown at the docs when enabled

**Reply by danhlee329** [reply]
Hmmmm, so this description found in the docs sort of makes sense: `When an array or object configuration option is mergeable, it means that values inside it will be added to any existing object or array that existed with the same name.`

By 'same name', is that referring to the name of configuration option (i.e., `packageRules`, `excludeDepNames`)?

Also, I've updated the config-merge-behavior that I'd like to see (below), but from reviewing the docs further, not sure that this is possible. Under [packageRules](https://docs.renovatebot.com/configuration-options/#packagerules), I do see that we can `extends` within one rule, but is there a way to merge individual `packageRules` based on an indentifier of sorts?  For example, per configs below, we have a package rule with `matchDatasources` set with one data source (`custom.local_module_tracker`).  In a consumer-repo, we want to extend that particular rule by providing a `excludePackageNames` list (no other adjustments).  Is this something that's currently possible in Renovate?

<details><summary>Example Config (Repo)</summary>

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [],
  "packageRules": [
    {
      "matchDatasources": ["custom.local_module_tracker"],
      "branchName": "renovate/local_module_tracker",
      "labels": ["renovate", "local_module_tracker"]
    }
  ],
  "hostRules": [],
  "requireConfig": "required",
  "platform": "github",
  "token": "***********"
}
```

</details>

<details><summary>Example Config Extended (App)</summary>

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": ["github>$ORG/$REPO//renovate.json"],
  "packageRules": [
    {
      "matchDatasources": ["custom.local_module_tracker"],
      "excludePackageNames": ["dep_name"]
    }
  ],
  "hostRules": [],
  "requireConfig": "required",
  "platform": "github",
  "token": "***********"
}
```

</details>

<details><summary>Example Config Extended (Final merged config)</summary>

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": ["github>$ORG/$REPO//renovate.json"],
  "packageRules": [
    {
      "matchDatasources": ["custom.local_module_tracker"],
      "branchName": "renovate/local_module_tracker",
      "labels": ["renovate", "local_module_tracker"],
      "excludePackageNames": ["dep_name"]
    }
  ],
  "hostRules": [],
  "requireConfig": "required",
  "platform": "github",
  "token": "***********"
}
```

</details>

**Reply by rarkins** [reply]
As a general rule, packageRules are not mergeable and we don't plan for them to be. 

If you have an unmet requirement please back a step up and state what you need to do in a high level description and not focus on how you're trying to do it (which won't work)

**Reply by danhlee329** [reply]
Thanks for the responses everyone, here's a high-level description of what we're trying to do:

`We would like to reuse and extend previously-defined package rules (from our config repo) in other app repos.  The main use-case for us is that we want to apply an 'exclude' list (based on dependency or package) in our app repos so that we can exclude specific deps/packages from being considered during the version checks.  This is assuming that the package rule in the config repo will consider all deps/packages that meet the matching conditions.`

While the use case is to exclude based on specifying for a package rule, I do see that there is a [ignoreDeps](https://docs.renovatebot.com/configuration-options/#ignoredeps) config which looks to be a catch-all.  I suppose this will be sufficient for us if the request above is considered overkill.

**Reply by rarkins** [reply]
I see in your original example you have this:

```
  "packageRules": [
    {
      "groupName": ["test-group-name"],
      "excludePackageNames": ["dep_name"]
    }
  ],
```

Ignoring that groupName must be a string and not an array (presumably it's just a typo), this rule could instead be:

```
  "packageRules": [
    {
      "groupName": null,
      "matchPackageNames": ["dep_name"]
    }
  ],
```
